src/StrMap.ts

```diff
@@ -295,32 +288,94 @@ export const pop = <A>(k: string, d: StrMap<A>): Option<[A, StrMap<A>]> => {
   return a.isNone() ? none : some(tuple(a.value, remove(k, d)))
 }
 
-const compact = <A>(fa: StrMap<Option<A>>): StrMap<A> => {
+const filterMap = <A, B>(fa: StrMap<A>, f: (a: A) => Option<B>): StrMap<B> => {
+  const value = fa.value
+  const result: { [key: string]: B } = {}
+  const keys = Object.keys(value)
+  for (const key of keys) {
```

Comment:
Isn't scanning twice here?

```ts
const keys = Object.keys(value)
for (const key of keys) {
}
```

what about?

```ts
for (const key in values) {
}
```

Reply:
The problem with `for ... in` is that it iterates over all keys even from prototype so either `Object.hasOwnProperty` check should be included or `Object.keys` should be used instead.

Reply:
@gcanti Search in fp-ts code shows 3 incorrect usages of `for ... in` in `Monoid`, `Semigroup` and `Setoid`. Should we fix them too?

Reply:
I always assumed `value` as a "simple" literal object (as it should be). In fact `for ... in` is used in many other functions in this module (`StrMap`)

Reply:
But typesystem can't restrict such value to always be a "simple" literal object so you can't be sure that you'll never get some additional keys from the prototype. 
```typescript
const numbers: {[key: string]: number} = {}
Object.setPrototypeOf(numbers, { evil: 'evil' })
for (const key in numbers) {
  const n = numbers[key]
  n.toFixed(2) // OOUCH!
}
```

I think the most safe and efficient way is to add `hasOwnProperty` everywhere, what do you think?

Reply:
There's even a tslint rule for such cases: https://palantir.github.io/tslint/rules/forin/

Reply:
On the other hand if you have an object with 10k keys set as prototype you end up with iterating over all that keys even if you object contains only one key. I think we should check what is faster - iterating over all keys with `hasOwnProperty` check or calling `Object.keys` and iterating over collection of keys. I think `V8` should perform some tricks to optimize calls to `Object.keys`, `Object.values` etc.

Reply:
Aaaaand the winner is: https://jsperf.com/object-keys-vs-hasownproperty/45

Reply:
LOL, ok let's add `"forin": true` to tslint.json